subprojects/launcher/src/main/java/org/gradle/launcher/daemon/client/DaemonClient.java

```diff
@@ -139,9 +139,13 @@ public BuildActionResult execute(BuildAction action, BuildActionParameters param
         LOGGER.debug("Executing build {} in daemon client {pid={}}", buildId, processEnvironment.maybeGetPid());
 
         int saneNumberOfAttempts = 100; //is it sane enough?
+        DaemonClientConnection newDaemon = null;
```

Comment:
I was thinking of something slightly different. Instead of carrying around `newDaemon` and figuring out what to do, we can duplicate the `Build build ...` lines.

Something like...
```
    public BuildActionResult execute(BuildAction action, BuildActionParameters parameters, BuildRequestContext requestContext) {
        UUID buildId = idGenerator.generateId();
        List<DaemonInitialConnectException> accumulatedExceptions = Lists.newArrayList();

        LOGGER.debug("Executing build {} in daemon client {pid={}}", buildId, processEnvironment.maybeGetPid());

        // Attempt to connect to an existing idle and compatible daemon
        int saneNumberOfAttempts = 100; //is it sane enough?
        for (int i = 1; i < saneNumberOfAttempts; i++) {
            DaemonClientConnection connection = connector.connect(compatibilitySpec);
            // No existing, compatible daemon is available to try, bail out.
            if (connection == null) {
                break;
            }
            // Compatible daemon was found, try it
            try {
                Build build = new Build(buildId, connection.getDaemon().getToken(), action, requestContext.getClient(), requestContext.getStartTime(), requestContext.isInteractive(), parameters);
                return executeBuild(build, connection, requestContext.getCancellationToken(), requestContext.getEventConsumer());
            } catch (DaemonInitialConnectException e) {
                // this exception means that we want to try again.
                LOGGER.debug("{}, Trying a different daemon...", e.getMessage());
                accumulatedExceptions.add(e);
            } finally {
                connection.stop();
            }
        }

        // No existing daemon was usable, so start a new one and try it once
        DaemonClientConnection connection = connector.startDaemon(compatibilitySpec);
        try {
            Build build = new Build(buildId, connection.getDaemon().getToken(), action, requestContext.getClient(), requestContext.getStartTime(), requestContext.isInteractive(), parameters);
            return executeBuild(build, connection, requestContext.getCancellationToken(), requestContext.getEventConsumer());
        } catch (DaemonInitialConnectException e) {
            // This means we could not connect to the daemon we just started.  fail and don't try again
            throw new NoUsableDaemonFoundException("A new daemon started but cannot be connected. " +
                    "This might indicate that the networking configuration blocks the connection to the daemon. " +
                    "Details of the created daemon: " +
                    "pid=" + connection.getDaemon() + ", " +
                    "address= " + connection.getDaemon().getAddress() + ". " +
                    "BuildActionParameters were " + parameters + ".", accumulatedExceptions);
        } finally {
            connection.stop();
        }
    }
```

Reply:
It's a bit too verbose to my taste, but easier to follow. I've adjusted the code.